smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/waiters/JMESPathVisitor.kt

```diff
@@ -0,0 +1,412 @@
+                    optionalityMark
+                ) {
+                    bufferWriter.playback()
+                }
+                val right = expression.right!!
+                return mappingBlock(right, filteredVar)
+            }
+            else -> throw Exception("Cannot filter non-list type: ${unfiltered.shape}")
+        }
+    }
+
+    // Returns the inner expression unchanged when the inner expression is an array of non-array elements.
+    // Returns the inner expression flattened when the inner expression is an array of arrays.
+    override fun visitFlatten(expression: FlattenExpression): Variable {
+        val toBeFlattened = expression.expression!!.accept(this)
+        return when (toBeFlattened.shape) {
+            is ListShape -> {
+                val elementShape = model.expectShape(toBeFlattened.shape.member.target)
+                when (elementShape) {
+                    is ListShape -> {
+                        // Double-nested List.  Perform Swift flat mapping.
+                        val flattenedVar = Variable("flattened", toBeFlattened.isOptional, elementShape)
+                        val dotOperator = "?.".takeIf { toBeFlattened.isOptional } ?: "."
+                        addTempVar(flattenedVar, "\$L\$LflatMap { $$0 }", toBeFlattened.name, dotOperator)
+                    }
+                    else -> {
+                        // Single nested List.  Return original list unchanged.
+                        toBeFlattened
+                    }
+                }
+            }
+            else -> {
+                toBeFlattened
+            }
+        }
+    }
+
+    // Implement contains() and length() free functions which are the only 2 JMESPath methods we support.
+    // contains() returns true if its 1st param is a collection that contains an element equal
+    // to the 2nd param, false otherwise.
+    // length() returns the number of elements of an array, the number of key/value pairs for a map,
+    // or the number of characters for a string.  Zero is returned if the argument is nil.
+    override fun visitFunction(expression: FunctionExpression): Variable {
+        when (expression.name) {
+            "contains" -> {
+                if (expression.arguments.size != 2) {
+                    throw Exception("Unexpected number of arguments to $expression")
+                }
+                val subject = expression.arguments[0]
+                val subjectVariable = subject.accept(this)
+                val search = expression.arguments[1]
+                val searchVariable = search.accept(this)
+                val optionalityMark = "?".takeIf { subjectVariable.isOptional } ?: ""
+                val returnValueVar = Variable("contains", false, boolShape)
+                return if (searchVariable.isOptional) {
+                    addTempVar(
+                        returnValueVar,
+                        "\$L.flatMap { \$L\$L.contains($$0) } ?? false",
```

Comment:
Why `flatMap` here? Should this just be `map` instead?

Reply:
This is the optional version of `flatMap`, not the array version.
https://developer.apple.com/documentation/swift/optional/flatmap(_:)#
Evaluates & returns the value of the closure if the receiver is non-nil, returns nil if the receiver is nil.

Reply:
right, so should we only be using `flatMap` if `subjectVariable` is optional and use `map` otherwise?

Reply:
No, map is not needed in the other case.
I am using Optional `flatMap()` as an easy, compact way to unwrap an optional for use in an expression, but default to `nil` if the optional is `nil`.
When the variable is not optional, I can just use it in the expression directly.

```swift
let result = subject.flatMap { collection.contains($0) } ?? false
```
is just a shorthand way of writing
```swift
let result: Bool
if let subject = subject {
    result = collection.contains(subject)
} else {
    result = false
}
```

Reply:
Exactly! `map` is what we should be reaching for in your specific example here!

The difference between `Optional.map` and `Optional.flatMap` is that the closures have different return types. 
`map` expects the closure to return a non-optional while `flatMap` expects the closure to return an optional. 

`Optional.map` -> `func map<U>(_ transform: (Wrapped) throws -> U) rethrows -> U?`
`Optional.flatMap` -> `func flatMap<U>(_ transform: (Wrapped) throws -> U?) rethrows -> U?`

So if we know the return type of the closure is non-optional, it is preferable to use map.

Reply:
Changed to `map`